### DISCUSSION TITLE
simplification and normalisation

### DISCUSSION THREAD

**Discussion by fclaudel** [bitwuzla/bitwuzla#141] [General]
Hi,
Bitwuzla class has simplify() which simplifies assert formulas.
Is there a way to simplify any Term created by mk_term?
e.g. given an EQUAL that takes 2 bitvector values, simplification on the term would result in either true or false
e.g. given an AND with multiple terms, existence of a false term would return false. True terms would be eliminated as unneeded.

Also, is there a way to normalize a term e.g. a given AND(a,b) and AND(b,a), return a normalized AND e.g. AND(a,b)

We want to use such simplification and normalization to implement a cache (a priori based on the term's hash) and avoid rerunning equivalent queries. Any help on this would be appreciated.

**Comment by aniemetz** [comment]
Currently we don't export such functionality, but this would be easy to add to the API. Will do and we'll keep you posted here.

**Reply by fclaudel** [reply]
Many thanks, please let me know when available (I'm happy to test a patch you send me)

**Comment by aniemetz** [comment]
@fclaudel I added support for simplifying terms to the API, see https://github.com/bitwuzla/bitwuzla/commit/f516ce42c79cd42aa7b51f52d1b262e1336a0642 on main.
